### DISCUSSION TITLE
Howto render embedded HTML tags from bokeh with jinja2 templates with fastAPI

### DISCUSSION THREAD

**Discussion by bijwaard** [fastapi/fastapi#9017] [Questions] [answered]
### First check

* [x ] I added a very descriptive title to this issue.
* [ x] I used the GitHub search to find a similar issue and didn't find it.
* [x ] I searched the FastAPI documentation, with the integrated search.
* [x ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case to get Bokeh graphs working with fastAPI (which does work with Flask).

```Python
from fastapi import FastAPI, Request
from fastapi.templating import Jinja2Templates
from bokeh.plotting import figure
from bokeh.embed import components
from bokeh.resources import INLINE

app = FastAPI()

templates = Jinja2Templates(directory="templates")

from jinja2 import Template

template = Template('''<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>LV Sensors</title>
        {{ js_resources }}
        {{ css_resources }}
    </head>
    <body>
    {{ plot_div }}
    {{ plot_script }}
    </body>
</html>
''')

@app.get("/graph")
async def read_root(request: Request, inline: bool=True):
    js_resources = INLINE.render_js()
    css_resources = INLINE.render_css()

    p = figure(title="my graph", x_axis_type='datetime', y_axis_type="linear")

    script, div = components(p, INLINE)
    if inline:
        return template.render(plot_script=script, plot_div=div, js_resources=js_resources, css_resources=css_resources)
    else:
        return templates.TemplateResponse("graph.html",
           { "request":request, "plot_script":script, "plot_div":div, 
             "js_resources":js_resources, "css_resources":css_resources } )

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app,host="localhost",port=8001)
```
Optional separate template file templates/graph.html used by templates.TemplateResponse looks like this:
```HTML
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>LV Sensors</title>
        {{ js_resources }}
        {{ css_resources }}
    </head>
    <body>
    {{ plot_div }}
    {{ plot_script }}
    </body>
</html>
```

### Description

* Open the browser and call the endpoint `/graph` or `/graph?inline=False`.
* It returns a very long HTML page with either literal HTML (when using template.render(..) or literal javascript in header and body starting with `&lt;script type=&#34;text/javascript&#34;&gt;` when using templates.TemplateResponse(...), i.e. it converted the < and > of HTML tags to something literal, such that the javascript is not working.
* But I expected it to return webpage with bokeh graph`, like below using Flask with the script below.
![image](https://user-images.githubusercontent.com/6523124/109163164-52de9880-7779-11eb-84e1-4c4e4aa570d3.png)

```Python
from flask import Flask, flash, render_template
from bokeh.plotting import figure, show
from bokeh.embed import components
from bokeh.resources import INLINE
from jinja2 import Template

template = Template('''<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>LV Sensors</title>
        {{ js_resources }}
        {{ css_resources }}
    </head>
    <body>
    {{ plot_div }}
    {{ plot_script }}
    </body>
</html>
''')

app = Flask(__name__,static_url_path = "")

@app.route("/graph")
def get_graph():
    js_resources = INLINE.render_js()
    css_resources = INLINE.render_css()

    p = figure(title="my graph", x_axis_type='datetime', y_axis_type="linear", plot_width=300, plot_height=300)

    script, div = components(p, INLINE)

    return template.render(plot_script=script, plot_div=div, js_resources=js_resources, css_resources=css_resources)

if __name__ == "__main__":
    app.run(debug=True, host="localhost",port=8001)
```

### Environment

* OS: Linux
* FastAPI Version: 0.63.0
* Python version:Python 3.8.5

**Comment by bijwaard** [comment]
Hi again,
I seem to have solved it with the inline template (i.e. inline=True), i.e. the approach that also works in Flask using HTMLResponse, i.e. in the following way:

```Python
...
from fastapi.responses import HTMLResponse

@app.get("/graph", response_class=HTMLResponse)
async def read_root(request: Request, inline: bool=True):
    js_resources = INLINE.render_js()
    css_resources = INLINE.render_css()
...
```
Is there a way this could work with `templates.TemplateResponse(...)`?

**Comment by falkben** [comment, marked as answer]
use the `|safe` jinja2 filter.

```jinja2
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8" />
        <title>LV Sensors</title>
        {{ js_resources|safe }} {{ css_resources|safe }}
    </head>
    <body>
        {{ plot_div|safe }} {{ plot_script|safe }}
    </body>
</html>
```

not clear to me without digging into jinja2 code as to why this is needed for the `TemplateResponse` and not `template.render()`

**Comment by bijwaard** [comment]
Hi falkben,

Thanks a lot, the |safe in the html file appears to help!

Kind regards,
Dennis

**Comment by gh4ag** [comment]
I also ran into this issue. Then, thanks to your hints, I went back to jinja2 v3.0 documentation ([link](https://jinja.palletsprojects.com/en/3.0.x/templates/)). I put it here if it can help solve this issue. I guess the culprits are the `\n` in the strings returned by Bokeh:

> When automatic escaping is enabled, everything is escaped by default except for values explicitly marked as safe. Variables and expressions can be marked as safe either in:

> a. The context dictionary by the application with markupsafe.Markup

> b. The template, with the |safe filter.

> If a string that you marked safe is passed through other Python code that doesn’t understand that mark, it may get lost. Be aware of when your data is marked safe and how it is processed before arriving at the template.

> If a value has been escaped but is not marked safe, auto-escaping will still take place and result in double-escaped characters. If you know you have data that is already safe but not marked, be sure to wrap it in Markup or use the |safe filter.

> Jinja functions (macros, super, self.BLOCKNAME) always return template data that is marked as safe.

> String literals in templates with automatic escaping are considered unsafe because native Python strings are not safe.
